diesel_derives/src/insertable.rs

```diff
@@ -26,33 +26,45 @@ pub fn derive(item: syn::DeriveInput) -> Result<proc_macro2::TokenStream, Diagno
     impl_generics.params.push(parse_quote!('insert));
     let (impl_generics, ..) = impl_generics.split_for_impl();
 
-    let (direct_field_ty, direct_field_assign): (Vec<_>, Vec<_>) = model
-        .fields()
-        .iter()
-        .map(|f| {
-            (
-                (field_ty(f, table_name, None)),
-                (field_expr(f, table_name, None)),
-            )
-        })
-        .unzip();
-
-    let (ref_field_ty, ref_field_assign): (Vec<_>, Vec<_>) = model
-        .fields()
-        .iter()
-        .map(|f| {
-            (
-                (field_ty(f, table_name, Some(quote!(&'insert)))),
-                (field_expr(f, table_name, Some(quote!(&)))),
-            )
-        })
-        .unzip();
+    let mut generate_borrowed_insert = true;
 
-    Ok(wrap_in_dummy_mod(quote! {
-        use diesel::insertable::Insertable;
-        use diesel::query_builder::UndecoratedInsertRecord;
-        use diesel::prelude::*;
+    let mut direct_field_ty = Vec::with_capacity(model.fields().len());
+    let mut direct_field_assign = Vec::with_capacity(model.fields().len());
+    let mut ref_field_ty = Vec::with_capacity(model.fields().len());
+    let mut ref_field_assign = Vec::with_capacity(model.fields().len());
+
+    for field in model.fields() {
+        let serialize_as = field.ty_for_serialize()?;
+        let embed = field.has_flag("embed");
+
+        match (serialize_as, embed) {
+            (None, true) => {
+                direct_field_ty.push(field_ty_embed(field, None));
+                direct_field_assign.push(field_expr_embed(field, None));
+                ref_field_ty.push(field_ty_embed(field, Some(quote!(&'insert))));
+                ref_field_assign.push(field_expr_embed(field, Some(quote!(&))));
+            }
+            (None, false) => {
+                direct_field_ty.push(field_ty(field, table_name, None));
+                direct_field_assign.push(field_expr(field, table_name, None));
+                ref_field_ty.push(field_ty(field, table_name, Some(quote!(&'insert))));
+                ref_field_assign.push(field_expr(field, table_name, Some(quote!(&))));
+            }
+            (Some(ty), false) => {
+                direct_field_ty.push(field_ty_serialize_as(field, table_name, &ty));
+                direct_field_assign.push(field_expr_serialize_as(field, table_name, &ty));
+
+                generate_borrowed_insert = false; // as soon as we hit one field with #[diesel(serialize_as)] there is no point in generating the impl of Insertable for borrowed structs
+            }
+            (Some(_), true) => {
+                return Err(field
```

Comment:
can we use here the span of the attribute only? I think that should give us a nicer error message.